### ISSUE TITLE
NaN after trained rbig transform

### ISSUE THREAD

**Issue by gzhu06** [IPL-UV/rbig#13]
First of all, thanks for the paper and code, it's very inspiring.

Specifically, first of all, I tried to generate 100 classes of vectors (each class contains 300 samples). These vectors follow multivariate student t distribution of 100 class dependent means and variances. Then I split these 100 classes into 80 and 20 for training and inferencing respectively. Finally, I try to use rbig following:

train_dataset_rbig = rbig_model.fit_transform(train_dataset)
test_dataset_rbig = rbig_model.transform(test_dataset)

then train_dataset_rbig is Okay, but NaN appears in test_dataset_rbig.

Am I wrongly using rbig or something else?

Thanks!

**Comment by jejjohnson** [comment]
Hi thanks for you interest! ￼

So in order to debug, would you be able to provide a minimal working code example? Then maybe I could see if there is an issue. My first thought might be a type issue but I'm not sure. Let me know!

**Comment by gzhu06** [comment]
import numpy as np
from sklearn import datasets
from rbig.rbig import RBIG, entropy
from toy_example import dict2array

import numpy as np

def dict2array(dicts):

	dataset = []
	labels = []

	for label in dicts:
		for data_sample in dicts[label]:
			
			dataset.append(data_sample)
			labels.append(int(label))

	return np.array(dataset), np.array(labels)

def multivariate_t(mu, sigma, dof, m):
    """
    Produce m samples of d-dimensional multivariate t distribution
    
    Args:
        mu (numpy.ndarray): mean vector
        sigma (numpy.ndarray): scale matrix (covariance)
        dof (float): degrees of freedom
        m (int): # of samples to produce

    Returns:
        numpy.ndarray
    """
    d = len(sigma)
    g = np.tile(np.random.gamma(dof / 2 , 2 / dof, m), (d, 1)).T
    z = np.random.multivariate_normal(np.zeros(d), sigma, m)
    return mu + z / np.sqrt(g)

def main():

	dist_type = 'Student-t' # Normal
	vec_dim = 128
	num_samples = 300
	num_classes = 100
	num_trials = 100000
	DoF = 3

	print('Generating dataset...')
	# generate Gaussian Prior on class
	C_y = datasets.make_spd_matrix(vec_dim)
	m_y = np.random.rand(vec_dim)

	y = np.random.multivariate_normal(m_y, C_y, num_classes)  # 50 classes in total

	# generate x based on class Gaussian Prior
	C_x = datasets.make_spd_matrix(vec_dim)
	x = {}
	x_train = {}
	x_test = {}

	for i, y_prior in enumerate(y):

		if i <  int(0.8*num_classes):
			if dist_type == 'Normal':
				x_train[str(i)] = np.random.multivariate_normal(y_prior, C_x, num_samples)
			elif dist_type == 'Student-t':
				x_train[str(i)] = multivariate_t(y_prior, C_x, DoF, num_samples)
 
		else:
			if dist_type == 'Normal':
				x_test[str(i)] = np.random.multivariate_normal(y_prior, C_x, num_samples)
			elif dist_type == 'Student-t':
				x_test[str(i)] = multivariate_t(y_prior, C_x, DoF, num_samples)

	print('Generating training dataset...')
	train_dataset, train_labels = dict2array(x_train)
	test_dataset, test_labels = dict2array(x_test)

	# rBIG
	n_layers = 1000
	rotation_type = 'pca'
	random_state = 123
	zero_tolerance = 10

	# Initialize RBIG class
	rbig_model = RBIG(n_layers=n_layers, rotation_type=rotation_type, random_state=random_state, zero_tolerance=zero_tolerance)

	# transform data
	print('RBIG Transform training...')
	train_dataset_rbig = rbig_model.fit_transform(train_dataset)
	print('RBIG Transform inferencing...')
	test_dataset_rbig = rbig_model.transform(test_dataset)

	print(np.isnan(np.sum(train_dataset_rbig)))
	print(np.isnan(np.sum(test_dataset_rbig)))


if __name__ == '__main__':
	main()

**Comment by gzhu06** [comment]
> 
import numpy as np
from sklearn import datasets
from rbig.rbig import RBIG, entropy
from toy_example import dict2array

import numpy as np

def dict2array(dicts):

	dataset = []
	labels = []

	for label in dicts:
		for data_sample in dicts[label]:
			
			dataset.append(data_sample)
			labels.append(int(label))

	return np.array(dataset), np.array(labels)

def multivariate_t(mu, sigma, dof, m):
    """
    Produce m samples of d-dimensional multivariate t distribution
    
    Args:
        mu (numpy.ndarray): mean vector
        sigma (numpy.ndarray): scale matrix (covariance)
        dof (float): degrees of freedom
        m (int): # of samples to produce

    Returns:
        numpy.ndarray
    """
    d = len(sigma)
    g = np.tile(np.random.gamma(dof / 2 , 2 / dof, m), (d, 1)).T
    z = np.random.multivariate_normal(np.zeros(d), sigma, m)
    return mu + z / np.sqrt(g)

def main():

	dist_type = 'Student-t' # Normal
	vec_dim = 128
	num_samples = 300
	num_classes = 100
	num_trials = 100000
	DoF = 3

	print('Generating dataset...')
	# generate Gaussian Prior on class
	C_y = datasets.make_spd_matrix(vec_dim)
	m_y = np.random.rand(vec_dim)

	y = np.random.multivariate_normal(m_y, C_y, num_classes)  # 50 classes in total

	# generate x based on class Gaussian Prior
	C_x = datasets.make_spd_matrix(vec_dim)
	x = {}
	x_train = {}
	x_test = {}

	for i, y_prior in enumerate(y):

		if i <  int(0.8*num_classes):
			if dist_type == 'Normal':
				x_train[str(i)] = np.random.multivariate_normal(y_prior, C_x, num_samples)
			elif dist_type == 'Student-t':
				x_train[str(i)] = multivariate_t(y_prior, C_x, DoF, num_samples)
 
		else:
			if dist_type == 'Normal':
				x_test[str(i)] = np.random.multivariate_normal(y_prior, C_x, num_samples)
			elif dist_type == 'Student-t':
				x_test[str(i)] = multivariate_t(y_prior, C_x, DoF, num_samples)

	print('Generating training dataset...')
	train_dataset, train_labels = dict2array(x_train)
	test_dataset, test_labels = dict2array(x_test)

	# rBIG
	n_layers = 1000
	rotation_type = 'pca'
	random_state = 123
	zero_tolerance = 10

	# Initialize RBIG class
	rbig_model = RBIG(n_layers=n_layers, rotation_type=rotation_type, random_state=random_state, zero_tolerance=zero_tolerance)

	# transform data
	print('RBIG Transform training...')
	train_dataset_rbig = rbig_model.fit_transform(train_dataset)
	print('RBIG Transform inferencing...')
	test_dataset_rbig = rbig_model.transform(test_dataset)

	print(np.isnan(np.sum(train_dataset_rbig)))
	print(np.isnan(np.sum(test_dataset_rbig)))


if __name__ == '__main__':
	main()
